api-server/api-server-common/src/storage/impls/postgres/mod.rs

```diff
@@ -32,6 +33,19 @@ use self::transactional::ApiServerPostgresTransactionalRw;
 
 pub struct TransactionalApiServerPostgresStorage {
     pool: Pool<PostgresConnectionManager<NoTls>>,
+    /// This task is responsible for rolling back failed RW/RO transactions, since closing connections are pooled
+    tx_dropper_joiner: tokio::task::JoinHandle<()>,
+    /// This channel is used to send transactions that are not manually rolled back to the tx_dropper task to roll them back
+    db_tx_conn_sender: tokio::sync::mpsc::UnboundedSender<
+        PooledConnection<'static, PostgresConnectionManager<NoTls>>,
+    >,
+}
+
+impl Drop for TransactionalApiServerPostgresStorage {
+    fn drop(&mut self) {
+        // Since the whole connection pool will be destroyed, we can safely abort all connections
+        self.tx_dropper_joiner.abort();
```

Comment:
Yes, but the lower level connection will be dropped since the connection pool will be destroyed, and hence the session will be dropped, and hence the transaction will be aborted anyway. At least this is my understanding of how postgres and mysql work from my research into this problem, but I'm open to being corrected here.

Reply:
Ok, if it's so that sounds acceptable, albeit still not the cleanest. If I understand it correctly, some transactions may not be explicitly rolled back by issuing the `ROLLBACK` query, relying on the connection cleanup procedure to effectively abort it. Is that about right?

Reply:
Yes, but I'm not sure whether there's a misunderstanding, so let me elaborate.

This is exclusively at the end of life of the database, which is probably obvious.

So, IF we're shutting down the system (hence the database object, not a transaction object) is about to be dropped, and there are connections that were transactions and are still not rolled back (even though their transaction objects are destroyed, by definition, because a transaction cannot outlive the database given the lifetime constraint with that phantom data), then aborting will, yes, cause the connection cleanup to effectively rollback the transaction, because the sessions of the connection pool will be ended, which cancels all low-level postgres transactions that were open. 

Having said that, again, this depends on my understanding of how database connections/sessions work and based my research on this problem. I'm happy to be challenged on that. 

Now the alternative is that we have to block when shutting down the database. I'm happy to do that if it turned out to be necessary... in fact it's like a line of code in tokio, but I consider it less clean because of my eternal fear of deadlocks.

Reply:
> Now the alternative is that we have to block when shutting down the database. I'm happy to do that if it turned out to be necessary... in fact it's like a line of code in tokio, but I consider it less clean because of my eternal fear of deadlocks.

That could be mitigated by using `spawn_blocking` for the cleanup task but then you won't be able to use the single-threaded tokio runtime. It's already the case in mempool because of the blocking subsystem handle business but it's not great and I'd like to avoid that if possible.